dashboard/k8s_utils.py

```diff
@@ -0,0 +1,122 @@
+import logging
+
+logger = logging.getLogger(__name__)
+
+CPU_SHARES_PATH = "/sys/fs/cgroup/cpu/cpu.shares"
+CPU_USAGE_PATH = "/sys/fs/cgroup/cpuacct/cpuacct.usage"
+PROC_STAT_PATH = "/proc/stat"
+
+container_num_cpus = None
+host_num_cpus = None
+
+last_cpu_usage = None
+last_system_usage = None
+
+
+def cpu_percent():
+    """Estimate CPU usage percent for Ray pod managed by Kubernetes
+    Operator.
+
+    Computed by the following steps
+       (1) Replicate the logic used by 'docker stats' cli command.
+           See https://github.com/docker/cli/blob/c0a6b1c7b30203fbc28cd619acb901a95a80e30e/cli/command/container/stats_helpers.go#L166.
+       (2) Divide by the number of CPUs available to the container, so that
+           e.g. full capacity use of 2 CPUs will read as 100%,
+           rather than 200%.
+
+    Step (1) above works by
+        dividing delta in cgroup's cpuacct.usage by
+        delta in total host cpu usage, averaged over host's cpus.
+
+    Since deltas are not initially available, return 0.0 on first call.
+    """ # noqa
+    global last_system_usage
+    global last_cpu_usage
+    try:
+        cpu_usage = _cpu_usage()
+        system_usage = _system_usage()
+        # Return 0.0 on first call.
+        if last_system_usage is None:
+            cpu_percent = 0.0
+        else:
+            cpu_delta = cpu_usage - last_cpu_usage
+            # "System time passed." (Typically close to clock time.)
+            system_delta = (
+                (system_usage - last_system_usage) / _host_num_cpus())
+
+            quotient = cpu_delta / system_delta
+            cpu_percent = round(quotient * 100 / container_cpu_count(), 1)
+        last_system_usage = system_usage
+        last_cpu_usage = cpu_usage
+        # Computed percentage might be slightly above 100%.
+        return min(cpu_percent, 100.0)
+    except Exception as e:
+        logger.exception("Error computing CPU usage of Ray Kubernetes pod.", e)
+        return 0.0
+
+
+def container_cpu_count():
```

Comment:
Should we be doing this in the core too? If so, can we move the cpu count part there?

Reply:
`ray.utils._get_docker_cpus` currently does min(A, B) with
A = quota/period
B = number of cpus indicated by cpuset.cpus
We could add
C = (cpu shares) / 1024
and do min(A, B, C)

What do you think?

Reply:
I would think that's preferable since I can't think of a case in which the dashboard should think the cluster has a different number of cpus than the scheduler.

Reply:
And probably after we do this, we'd want to use `ray.utils.get_num_cpus()` in `ReporterAgent` rather than `psutil.cpu_count`.

But we should probably respect what the tuple in this line is doing: 
https://github.com/ray-project/ray/blob/153dcd3734719d68b0ab2c9c0d4bec1fa85eaf87/dashboard/modules/reporter/reporter_agent.py#L116-L117
Does anyone know what this pair of normal count and logical count mean, 
and what it should mean in the context of a container running in a K8s pod?
@fyrestone

Reply:
For the immediate moment I'll avoid dealing with it by keeping the `if IF_IN_KUBERNETES`

Reply:
> And probably after we do this, we'd want to use `ray.utils.get_num_cpus()` in `ReporterAgent` rather than `psutil.cpu_count`.
> 
> But we should probably respect what the tuple in this line is doing:
> https://github.com/ray-project/ray/blob/153dcd3734719d68b0ab2c9c0d4bec1fa85eaf87/dashboard/modules/reporter/reporter_agent.py#L116-L117
> 
> 
> Does anyone know what this pair of normal count and logical count mean,
> and what it should mean in the context of a container running in a K8s pod?
> cc @fyrestone

Sorry, I am not familiar with K8s resources. The normal count and logical count is defined here: https://psutil.readthedocs.io/en/latest/#psutil.cpu_count